### PR TITLE
fix(router): handle non-trailing slashes correctly

### PR DESCRIPTION
**Pull request by hekike**
## Issues

Closes:

https://github.com/restify/node-restify/issues/1464

Non-trailing slash leads to incorrect routing match (and parameter parsing).

# Changes

Fix router regexp matching with non-trailing slash(es)

### CODE REVIEW

File: test/router.test.js [hunk 1]
```diff
@@ -210,17 +210,30 @@ test('Default non-strict routing ignores trailing slash(es)', function (t) {
 
     var trailing = server.router.routes.GET[0];
     t.ok(trailing.path.test('/trailing/'));
-    t.ok(trailing.path.test('//trailing//'));
+    t.ok(trailing.path.test('/trailing//'));
     t.ok(trailing.path.test('/trailing'));
 
     var noTrailing = server.router.routes.GET[1];
     t.ok(noTrailing.path.test('/no-trailing'));
-    t.ok(noTrailing.path.test('//no-trailing//'));
```
Comment (retrohacker) [inline test/router.test.js]: I'm not sure I have enough context to understand why this change is necessary. Why did we change an existing test as opposed to adding a separate test for a single leading `/`?

Comment (hekike) [reply test/router.test.js]: The original test case is not related to the code what it tests. This test should focus to double no-trailing slashes.

Comment (hekike) [reply test/router.test.js]: Sry, I mean it should focus to slashes at the end of the path :P

**Comment by DonutEspresso** [comment]
@hekike thanks for taking this on! I think you're right, the behavior should be to not remove double slashes by default - that should be taken care of by the dedupeSlashes plugin. This is going to be a breaking change, perhaps we can slide this in before we cut 6.x @retrohacker?

**Comment by hekike** [comment]
Apologize I didn't explain very well the context of this pr.
To be on the same page, the current PR:

- Allows only one slash at the middle to prevent route pram parsing issues
- Allows **zero or more** slashes at the end

I believe what @DonutEspresso proposes is a one step further and wouldn't allow, more than one slash at the end neither.

The new behavior after this proposal would be:

- Allows only one slash at the middle to prevent route pram parsing issues
- Allows **zero or one** slash at the end

**Comment by retrohacker** [comment]
Thanks for tackling this @hekike! 

I'm :+1: on leaving the path exactly as is, and allowing the user to specify the dedupe behavior through a plugin. My main motivation is that we should do as little "magic" as possible, and changing paths before they get to the user seems like magic. I think our plugin logic should behave like [`merge_slashes` in nginx](http://nginx.org/en/docs/http/ngx_http_core_module.html#merge_slashes). If users need a different behavior, they can specify that as their own plugin.

**Comment by retrohacker** [comment]
So, tl;dr, I propose:

1. We remove all url scrubbing logic from the router
2. We update our dedupe plugin to match NGinx's behavior
3. We document how you can specify your own url transformer through `server.pre`.

**Comment by retrohacker** [comment]
As for holding v6, I'd suggest we go ahead and ship. We've been in a state of perpetually staged breaking changes before (as @yunong pointed out yesterday) and it wasn't a great place to be.

Instead, I'd suggest we create a new branch called `next` that we merge `BREAKING CHANGE`s into until we think we have enough for a new major release. Then we can cut from there. This gives us two advantages:
1. We don't have to block bug/feature fixes because `BREAKING` is on the current `master` (like we are doing now)
2. Users who need the new `BREAKING` changes can opt into tracking the `next` branch by using `restify/node-restify#next` in their `package.json`.

**Comment by hekike** [comment]
@retrohacker I removed the double slash merging magic from the router, see my latest commit.
The `restify.plugins.pre.dedupeSlashes()` plugin is already documented.

I also refactored the `dedupeSlashes` test to cover cases with more than two slashes.
If I understand correctly `restify`'s dedupleSlashes plugin already worked like `nginx`'s `merge_slashes` option, just the test didn't point it out earlier:

> Enables or disables compression of two or more adjacent slashes in a URI into a single slash. - [merge_slashes](http://nginx.org/en/docs/http/ngx_http_core_module.html#merge_slashes)

File: lib/router.js [hunk 2]
```diff
@@ -103,7 +103,7 @@ function compileURL(options) {
             return (false);
         }
 
-        pattern += '\\/+';
+        pattern += '\\/{1}';
```
Comment (retrohacker) [inline lib/router.js]: I think this quantifier is unnecessary

File: lib/router.js [hunk 3]
```diff
@@ -134,7 +134,7 @@ function compileURL(options) {
     }
 
     if (!options.strict) {
-        pattern += '[\\/]*';
+        pattern += '[\\/]?';
```
Comment (retrohacker) [inline lib/router.js]: I think this list is unnecessary

**Comment by retrohacker** [comment]
This shouldn't merge into `master` unless we want to cut a new release since it is going to be a breaking change. Can we open this PR against a new branch `next`?

**Comment by hekike** [comment]
@retrohacker I've no permission to create a `next` branch in this repo, could you create one? I cannot open a PR to a non-existent branch.

**Comment by hekike** [comment]
Closed by the favor of https://github.com/restify/node-restify/pull/1512
cc @retrohacker
